src/kOS/Execution/Stack.cs

```diff
@@ -48,9 +52,19 @@ private void ThrowIfInvalid(object item)
         /// <returns></returns>
         private object ProcessItem(object item)
         {
-            if (item is float)
+            if (item is float) {
                 // promote floats to doubles
-                return Convert.ToDouble(item);
+                item = Convert.ToDouble (item);
+            }
+            if (item is double)
+            {
+                if (!Double.IsNaN((double)item) && Int32.MinValue < (double)item && (double)item < Int32.MaxValue)
+                {
+                    int intPart = Convert.ToInt32(item);
+                    if ((double)item == (double)intPart)
+                        item = intPart;
+                }
+            }
```

Comment:
@abenkovskii You are asking the wrong question.  The question should be "show me one place where a cast to a double exists just in case the value is an int", rather than trying to find a place where a bug exists because that cast has failed to be done and therefore the code is currently already broken.

There are places that expect a double, but they won't break when given an int precisely because that casting I'm referring to is in the code.  My point is that regardless of whether you pick a standard that the stack will always turn things into integers or the pick a standard that it will always turn things into doubles, either way there will have to be places where that standard isn't what's needed and they'll need to do a cast from that back to what they need..  Places like that already exist for making ints into floating point (for example, when you ``lock throttle to 1`` instead of ``lock throttle to 1.0``, kOS's flybywire logic will convert that integer 1 into a floating point 1.0 for you.  The stack isn't doing it - it's an exception that had to be hardcoded for that one case.  The point I'm making is that either you default to using ints for all values, in which case you cast the ints to floating point where needed in C#'s code, or you default to using doubles for all values, in which case you cast the doubles to ints where needed in kOS's C# code.  Either way the standard doesn't solve all your casting needs on its own because whichever standard you pick, there will be parts of the code that have to convert back out of it again - so maybe the right answer is to assume neither and have the casting everywhere that it matters to the application at hand, which is pretty much how it works right now except that this bug is that the casting was missing in one spot where it was needed.  That's no more or less a hole in the design than the fact that it's possible in the future to forget a cast to a double where it is needed.

That's why I'm beginning to think doing this at a low level of the stack isn't the right place for it - unless we make a Number wrapper class like we're planning on making a String wrapper class.
Then the Number wrapper class would have a GetAsInt or GetAsDouble or GetAsFloat, depending on what the underlying kOS code needs to do with it, and you'd make it so that it always has to pick one of those (so that you can't forget to cast).